### ISSUE TITLE
feature: HTTPS based deploy keys

### ISSUE THREAD

**Issue by TomFreudenberg** [go-gitea/gitea#2051]
## Description

Hi, I have found at https://gitlab.com/gitlab-org/gitlab-ce/issues/20845 a really senseful feature request and would like to ask for something similar at gitea.

> We have a client service that needs to clone a repository. It cannot use ssh, and thus cannot use a regular ssh-based deploy key.
Is it possible to create a HTTP(S)-based deploy key, so that we can configure the client to access https://SECRETKEY@gitlaburl.com/repo.git?

The repo can be accessed like this:

```
git clone https://gitlab-ci-token:<YOUR_TOKEN>@<GITLAB_INSTANCE_URL>/<USERNAME_OR_GROUPNAME/<REPO_NAME>.git
```

### Customer Proposal

Would it not be smart to extend the "deploy keys" concept with HTTPS-based keys, like this:

* Create the deploy key directly in the user interface (no need to run ssh-keygen locally) 
* Works like the ssh-based deploy key, but for https 
* (?) Share it between projects, for easy managability 
* Create multiple keys for multiple clients, and revoke the ones you need

Thanks for your work and feedback

Tom

**Comment by silverwind** [comment]
Seconding this request. It would be quite handy if there were per-repo access tokens which allow read or read/write access to a single repo via HTTPS, so basically the same functionality like deploy keys, but over HTTPS.

**Comment by kgraefe** [comment]
I think it would be nice to have both per-account and per-organization access tokens.

**Comment by DjSni** [comment]
> I think it would be nice to have both per-account and per-organization access tokens.

i agree with you, but you should be able to set for each token which rights the token has. 
For example, only read rights, read and write rights...
But I hope this will be implemented soon.

**Comment by Snck3rs** [comment]
would be a great feature! i hope someone will be found to implement that.
Great work, btw!

**Comment by multicast** [comment]
https based deploy key per organization would also be interesting, as well as ssh deploy key per organization would simplify many setups.
currently I had to create "technical" user to achieve this and make it a member of the team/organization. The organization admin should be able to generate such tokens or mark keys as deploy keys for organization or multiple repositories.

**Comment by skorokithakis** [comment]
Yes please, HTTPS cloning with tokens would be great for times where adding an SSH key is inconvenient.

**Comment by vbichov** [comment]
this issue is open for almost 5 years - what are the chances this will actually be implemented?

**Comment by lunny** [comment]
I think it's not very difficult, if the token is a user's personal access token. We need to change here https://github.com/go-gitea/gitea/blob/main/routers/web/repo/http.go#L113 and create a new implementation of https://github.com/go-gitea/gitea/blob/34b5436ae1af2735546bb519a950eabf4990212d/services/auth/interface.go#L23 only for SmartHTTP routers.

**Comment by silverwind** [comment]
It should generate a unique token per-repo, not re-use a user's personal token. But I guess the implementation can probably share a lot of code with the personal token mechanism.

**Comment by skorokithakis** [comment]
@lunny I disagree, sharing one key for multiple repositories is the job of the personal token. A deployment key is bound to a single repo, and you may have multiple keys in one repo.

**Comment by TomFreudenberg** [comment]
I agree to @skorokithakis 

If you will use such deploy keys - I guess you will use https - so you will deploy your repo on several instances - each instance could have its own key per repo

**Comment by multicast** [comment]
> 
> > Share it between projects
> 
> One key could be used for multiple repositories, one repositories could also have multiple keys.

I fully agree. One https token (or ssh key) should be usable in more repositories, and one repository should be able to have more tokens (ssh keys). Some of them read-only, some read-write. Also a per-organization tokens (keys) should exist. Now I have to solve it via some bot account, even this is not enough often.

Currently I can have an ssh deploy key in a repository (and enable write access if necessary), so my production servers can use the ssh key when pulling from host e.g. `git.corp`. Since the deploy key cannot be reused in other repo, I am forced to use the ssh config file:
```
Host foo.git.corp
  IdentityFile ~/.ssh/foo
Host bar.git.corp
  IdentityFile ~/.ssh/bar
Host *.git.corp
  Hostname git.corp
  User git
```
If the project is composed from dozens of repos, I need dozens keys, one for every new repo. Reusing one deploy key for several repos would be a valuable feature, and the same applies for the https token.
If all  repos for particular project are grouped in an organization, this would make use case for the per-organization deploy key (or  https token).
Replicating the production setup to run independently an integration, q&a, or a test environment makes perfect sense not to share ssh key with production. That's why more deploy keys are already possible per repo (and so more https tokens must be possible too).

I do not use above setup any more - instead of managing multiple keys, on multiple servers, in multiple environments, I have a bot and that bot is an organization member or collaborator. It's enough to use bot's private key instead of mine or managing dozens of keys. Replacing this with single https token would make things incredibly simpler.

Do not limit an https token (or a deploy key) to single repo. It should work, potentially with different privilege level, for multiple other repos. If somebody "overloads" the key or token with access to a too-many repos, the key or token has more "power", but it means you can also revoke/invalidate a key or token from all repos efficiently.

**Comment by techknowlogick** [comment]
Tbh I use bot accounts for this so I can assign exactly which repos/orgs they have access to.

**Comment by multicast** [comment]
Me too, but sometimes you I have a "3rd party" for single or few closely related repos, for which the https token is the only way (i.e. no ssh access, so no deploy keys). having a thousands 3rd parties means thousands of bot accounts versus simple https token. The 3rd party can be a single server or a cluster, not actualy a human, obviously.

**Comment by multicast** [comment]
I just found a bot account issue #13044, which helps using bot accounts, but is still not a replacement for https tokens.

**Comment by silverwind** [comment]
Yes, I also use "bot" style accounts to share one "deploy token" for multiple repos. But sharing such tokens on multiple systems is not a good practice security-wise because if one system gets compromised, more data is at risk than it needs to be.

So I would even prefer if a HTTPs deploy token is only valid for a single repo, just like SSH deploy keys currently are.

**Comment by MrSuicideParrot** [comment]
Bot accounts are a hacky way to minimise this problem and not the solution.
This issue has been open for 4 years and in my point of view, this is the only major feature that is missing when compared with gitlab and GitHub. 
Thus, I don't understand why this feature isn't a priority...

**Comment by lunny** [comment]
Vote:
- 💯 One key one repo
- 👍 One key could be used for multiple repositories, one repositories could also have multiple keys.
- 🎉 Just keep to use bot account

Or if you have any other options, please let me know.

**Comment by mgax** [comment]
> Or if you have any other options, please let me know.

GitHub allows you to set multiple deployment keys per repo, but you're not allowed to reuse a key on multiple repos (which is a royal pain). Also, GitHub only allows SSH deployment keys, it would be great to have HTTPS deployment keys too. So, for completeness sake, there can be another option, _one repository can have multiple keys but they can't be reused on another repo_. Not that I'm advocating for it :)

**Comment by lunny** [comment]
> > Or if you have any other options, please let me know.
> 
> GitHub allows you to set multiple deployment keys per repo, but you're not allowed to reuse a key on multiple repos (which is a royal pain). Also, GitHub only allows SSH deployment keys, it would be great to have HTTPS deployment keys too. So, for completeness sake, there can be another option, _one repository can have multiple keys but they can't be reused on another repo_. Not that I'm advocating for it :)

Added as a new vote option.

**Comment by DjSni** [comment]
Would like to set a 🥇 but i can´t.  I can only set 👍, 👎,😄,🎉,😕,❤️,🚀,👀
Thanks

**Comment by lunny** [comment]
> Would like to set a 🥇 but i can´t. I can only set 👍, 👎,😄,🎉,😕,❤️,🚀,👀 Thanks

Changed to ❤️

**Comment by jerrykan** [comment]
> > Would like to set a 1st_place_medal but i can´t. I can only set +1, -1,smile,tada,confused,heart,rocket,eyes Thanks
> 
> Changed to heart

what about :100: ?

**Comment by silverwind** [comment]
Can't vote 💯, but it's definitely my favorite. Keep it simple and secure, at least initially. More flexibility can be added later and possibly also be added for SSH-based keys.

**Comment by lunny** [comment]
@jerrykan @silverwind Don't know why I can. Changed to 🚀

**Comment by multicast** [comment]
@silverwind - I fully agree with security-wise practice of not sharing keys. but one key/token per repo forces me to share key on multiple clients. so, security-wise, repo must have possibility of more keys.

on the other side, sharing single key/token for multiple repos puts them all at the same risk when a key/token is compromised. but possibility to reuse the key/token in multiple repos does not require you to do so. you can still follow policy of not reusing the key. but where appropriate, greatly simplifies setups.

@techknowlogick - I also agree that a bot account is workaround, not a solution.

**Comment by benedictjohannes** [comment]
Is there any fork already working on this?

**Comment by manuelchichi** [comment]
Any update on this feature?

**Comment by netthier** [comment]
This would also be useful for fine-grained control over the Gitea container registry, as it does not support SSH authentication, only HTTP(S).

**Comment by cwchristerw** [comment]
This is maybe related to https://github.com/go-gitea/gitea/pull/24767

**Comment by Jamesits** [comment]
Fine-grained API tokens would be great. Rather than having deploy keys shared between projects, would it be better to just have bot accounts that are PAT&SSH keys only, no UI login, no SSO? This allows sharing keys between projects. We also want different username and profile pictures for these bot accounts, so that CI released packages does not say they are uploaded by some human.

**Comment by HammyHavoc** [comment]
Is anyone currently working on this? Just hit this problem and might be interested in taking it on. Any rough stuff or notes anyone has about it would be appreciated.

**Comment by multicast** [comment]
I am not, even I miss this feature. Is it worth to summarize the target? Or settle on the winner - https://github.com/go-gitea/gitea/issues/2051#issuecomment-985647044

**Comment by lesinigo** [comment]
IMHO sharing the same project deploy key between multiple use cases should be avoided at all costs for security reasons (as @multicast also said). This brings us to the requirement of supporting more than one key per project.

Personally I'm against sharing the same "key" between multiple projects, I'd prefer the "bot account" approach, that would also probably be easier to implement since the logic and UI to manage which projects can be accessed by which users is already there.

Still, simple "deploy tokens" for projects, with read only access and with the ability to create more than one for each project, would probably solve 90% of use cases out there and can co-exist with a future "bot account" implementation. [Gitlab does the same too](https://docs.gitlab.com/ee/user/project/deploy_tokens/), for example.

**Comment by multicast** [comment]
Bot accounts are already possible and are certainly workaround
- they must be specialy secured (password, key, api tokens...)
- they steal the first level of namespace used by users or organizations
- they clutter the list of instance users, members of the repository, organization or team
- to change their public key or token, you must login to them
- to create account a developer needs enough privileges (or bother admin)
- overtime, nobody remembers who created or "owns" the bot12345 and if it can be deleted
- managing bot's properties via command line or API needs bot's own credentials
- accounts can be in an external corporate system (SAML, LDAP...) the gitea admin has no easy control of.

Tokens or keys:
- your (developer) account is probably already secured
- no additional namespace pollution or any user lists with new entry
- changing your keys can be done yourself
- removing repository or the organization also removes key/token association
- should be naturally managed using API or tea command line (with your credentials)

Using the same "key" for multiple projects is not sharing. It's empowering the key for having access to more, often tightly related repositories (e.g. subtree or submodules). Allowing to reuse key or token does not block your preferred approach - feel free to use any key or token only once, have different per each repo.

Preventing reuse of a repo-based deploy public ssh key from accessing more than one repo, or blocking a token access to multiple repositories, implies creation of many keys or tokens and their management on clients accessing Gitea. Allowing single powerful token to access multiple repositories also allows (e.g.  after a compromise) to revoke, delete or update single key and effectively block access of the old token wherever it ever had access to.

The winner says it all: _One key could be used for multiple repositories, one repository could also have multiple keys_
